### PR TITLE
Reimplement WBPn Calculation Procedure

### PR DESCRIPTION
**Pull request by bska**
This PR implements the `WPAVE` keyword and its associate WBPn well level report quantities.  We support all valid settings of `WPAVE` and `WWPAVE`.  The various weighted averages are accumulated through compensated summation of both the numerator and denominator terms for the inner block, the direct/immediate level 1 neighbours, and the diagonal level 2 neighbours.  We combine those contributions to the `WBP`, `WBP4`, `WBP5`, and `WBP9` terms when we "commit" the contributions (`Accumulator::commitContributions()`), per connection for values weighted by the connection transmissibility factor, and per well for values weighted by pore volumes.

We distinguish `OPEN` from `ALL` connections through callback functions in the implementation of `accumulateLocalContributions()` and the per connection weighting terms are provided as another set of callback functions depending on the value of the inner block weighting factor F1.

Depth correction (`NONE`, `WELL`, or `RES`) is affected through a separate set of callback functions used in the implementation of `connectionPressureOffset()`.

We discover source locations in a two-step process.  At WBPn calculation construction time, we exclude only those cells that are outside the model's dimensions.  The user is then expected to call member function `pruneInactiveWBPCells()`, typically at the `CalculatorCollection` level, at a later time in order to prune inactive cells.  This two-step split is necessary because we do not have a complete global view of the model's active cells in a parallel run.  It is very possible that the WBPn calculation on one rank will require source values from another rank and that begets this extra caution.

The user must call `inferBlockAveragePressures()` to compute the WBPn values.  Once completed, the result values may be extracted by calling member function `averagePressures()`.  We expect that the user will provide a complete set of up-to-date source values when calling this member function, both at the reservoir and the well connection levels.

### CODE REVIEW

**Comment by bska** [comment]
I'm creating this PR in "draft" state because it depends on, and contains, PR #3558.  I will keep the PR in draft state until such time as it is ready for review.

**Comment by akva2** [comment]
how about you split off https://github.com/OPM/opm-common/pull/3559/commits/1a2182e4f8f6f840fca9e6c857b3cd382cc03425

File: opm/input/eclipse/Schedule/Well/PAvgCalculator.hpp [hunk 1]
```diff
@@ -17,93 +17,787 @@
...
+        /// Number of block-averaged pressure kinds/modes.
+        static constexpr auto NumModes =
+            static_cast<std::size_t>(WBPMode::WBP9) + 1;
 
-    struct Neighbour {
-        Neighbour(double porv_arg, std::size_t index_arg) :
-            porv(porv_arg),
-            global_index(index_arg)
-        {}
+        /// Storage type for block-avegraged well pressure results.
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Well/PAvgCalculator.hpp:91]: averaged

File: opm/input/eclipse/Schedule/Well/PAvgCalculator.hpp [hunk 2]
```diff
@@ -17,93 +17,787 @@
...
+    ///
+    /// \param[in] connections List of reservoir connections for single
+    ///   well.
+    PAvgCalculator(const GridDims&        cellIndexMap,
+                   const WellConnections& connections);
+
+    /// Destructor.
+    virtual ~PAvgCalculator();
+
+    /// Finsih construction by pruning inactive cells.
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Well/PAvgCalculator.hpp:177]: Finish

File: opm/input/eclipse/Schedule/Well/PAvgCalculator.hpp [hunk 3]
```diff
@@ -17,93 +17,787 @@
...
+
+    /// Top-level entry point for computing the pressure correction
+    /// term/offset of each active reservoir connection
+    ///
+    /// Will dispatch to lower level calculation routines based on algorithm
+    /// selection parameter in procedure controls.
+    ///
+    /// \param[in] sources Connection and cell-level raw data.
+    ///
+    /// \param[in] controls Averging procedure controls.  This function uses
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Well/PAvgCalculator.hpp:763]: Averaging

**Review by akva2 — COMMENTED** [review]
Some spotted doxy typos

**Comment by bska** [comment]
> how about you split off 1a2182e

Good idea.  I've created #3579 to do that.

**Review by akva2 — APPROVED** [review]
I've gone through the code and nothing jumps out at me. Admittedly it's not the smallest of changes and some stuff may have gone me by, but I think this can be merged since test coverage seems quite good.

**Comment by bska** [comment]
> I've gone through the code and nothing jumps out at me

Thanks.  I've corrected the misprints you identified earlier and I'll mark it as "ready for review" just to enable merging.

> Admittedly it's not the smallest of changes

It absolutely is not.  The biggest problem is that the original version completely ignore the issue of depth correction which has a profound impact on the logic.

> and some stuff may have gone me by, but I think this can be merged since test coverage seems quite good.

Thanks, I did try to cover my bases.  That said, I only really tested "reasonable" property values, so it make break down if someone passes negative pore-volumes or pressures and/or zero mixture densities.  I **believe** the latter is covered by logic in the weighted running average class though.

**Comment by bska** [comment]
jenkins build this please

**Comment by blattms** [comment]
As this is reviewed and approved by @akva2, I guess we should merge this?
Just one more jenkins run ... Feel free to self-merge...

**Comment by blattms** [comment]
jenkins build this please

**Comment by bska** [comment]
> Just one more jenkins run ... Feel free to self-merge...

Thanks.  Build check is green so I'll merge into master.

**Pull request merged**